Extract an IP address from CIDR reported by Qemu guest

mfranczy opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
qemu-guest-agent reports ip addresses of VMI with mask, like x.x.x.x/mask, this breaks 3rd party components which expect to get IP address in x.x.x.x format. This is visible when you use multus and have qemu-guest-agent installed.

We treat VMs created by KubeVirt as worker nodes for another k8s cluster and during network initialization we realized that calico has a problem with issue described above, for instance:
```
Error from server: failed to construct request for https://10.x.x.106%2F32:10250/containerLogs/kube-system/calico-node-lvcgj/calico-node, got parse https://10.225.128.106%2F32:10250/containerLogs/kube-system/calico-node-lvcgj/calico-node: invalid URL escape "%2F"
```

for more details look at the issue attached below:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
https://github.com/gardener/gardener-extension-provider-kubevirt/issues/33

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Set only an IP address for interfaces reported by qemu-guest-agent. Previously that was CIDR.
```

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -605,6 +605,14 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
+
+					// Qemu guest reports IP addresses in x.x.x.x/mask format
+					// remove the mask from the ip address
+					interfaceStatus.Ip = strings.Split(interfaceStatus.Ip, "/")[0]
```
Would `net.ParseCIDR("192.0.2.1/24")` help?

davidvossel replied:
also, can we get a unit test so we don't goof this up in the future somehow?

mfranczy replied:
@rmohr yes that's better. @davidvossel sure I will add unit tests as well.

rmohr replied:
Since we don't own the guest-agent, maybe we should also be cautious here and check if it is with or without subnet provided.


To illustrate the flow:
```
if err := net.ParseCIDR(); err != nil { # with netmask
   if err := net.ParseIP(); err != nil { # without netmask
     log.Error("qemu-guest-agent sent garbage")
     continue # not failing, but rather skipping this ip.
  }
}

```

mfranczy replied:
I see that `ParseCIDR` checks if IP address is correct https://golang.org/src/net/ip.go?s=16533:16577#L702 so I think it's not needed to run `ParseIP` again.

I updated tests, let's see if everything passes.

rmohr replied:
It may check if the IP is correct, but it fails if no CIDR is included: https://play.golang.org/p/sSkJYZ3JFDR

mfranczy replied:
yeah I realized that when I tested things :) however it's solved as we spoke here https://github.com/kubevirt/kubevirt/pull/4217#discussion_r496522696

rmohr reviewed (commented):
Nice catch!

mfranczy commented:
I see failing unit and func tests, will take a look

davidvossel commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1916,8 +1916,11 @@ var _ = Describe("VirtualMachineInstance", func() {
-			ip := "2.2.2.2/24"
-			ips := []string{"2.2.2.2/24", "3.3.3.3/16"}
+			cidr := "2.2.2.2/24"
+			cidrs := []string{"2.2.2.2/24", "fe80::3825:65ff:fe6e:1ee9/64"}
```
for completion, can we add an IP to this list that doesn't have a mask as well? 

Your logic checks out as it is now, this request is to ensure we future proof this a bit if someone else touches the logic.

mfranczy replied:
good point, thanks for review, I will do as you suggest.

davidvossel reviewed (commented):
looks great. I just made one super minor suggestion.

rmohr commented on pkg/util/net/ip/ip.go:
```diff
@@ -82,3 +85,28 @@ func isIPv6Disabled(filename string) bool {
+func ChangeCIDRsToIPs(ifs *api.InterfaceStatus) error {
+	// if IP address was the passed value we should do nothing
+	if net.ParseIP(ifs.Ip) == nil {
+		ip, _, err := net.ParseCIDR(ifs.Ip)
```
I would still prefer, like in my initial comment (and david hints it), that if `net.ParseCIRD` fails (it will fail on pure IP addresses without cidr, we should also try to do `net.Parse`.

mfranczy replied:
ok, I will do it as you ask. Soon I will update the PR, thanks for review.

mfranczy replied:
but just to highlight one thing, I do check `net.ParseIP` before `net.ParseCIDR`, if CIDR (or anything that is not an IP address) is passed to `net.ParseIP` the function will return `nil` (check here https://github.com/kubevirt/kubevirt/pull/4217/files/05eeda4aafcf6c9bc0788da1057a9d69389d8324#diff-4713d8334b0e7a78d689e3a725ab753fR91), and then I check if this is CIDR. So it's the same as you say but in a different order. However I will do it as you wish.

rmohr replied:
If you do that, fine, then I just misseed it.

rmohr reviewed (changes requested):
Very close. We should handle the non-cidr case, like David said.

phoracek commented:
This breaks all the existing users who rely on the CIDR.

/hold

/cc @qinqon @AlonaKaplan @EdDev @maiqueb

rmohr commented:
> This breaks all existing users who rely on the CIDR.

So you say that we have mixed reports, with and without CIRD, is a feature?

qinqon commented on pkg/util/net/ip/ip.go:
```diff
@@ -82,3 +85,28 @@ func isIPv6Disabled(filename string) bool {
+	}
+
+	for i, v := range ifs.IPs {
+		if net.ParseIP(v) == nil {
+			ip, _, err := net.ParseCIDR(v)
+			if err != nil {
+				return fmt.Errorf("qemu-guest-agent reported invalid CIDR value '%s' for '%s': %v",
+					v, ifs.InterfaceName, err)
+			}
+			ifs.IPs[i] = ip.String()
+		}
```
Can we wrap this at a helper function (now that we even add a fallback to net.ParseIP)

phoracek commented:
> > This breaks all existing users who rely on the CIDR.
> 
> So you say that we have mixed reports, with and without CIRD, is a feature?

No, I haven't said that.

I'm saying that moving in either direction is breaking users who assumed one or the other is a feature. Removing CIDR everywhere means that we would be unable to figure out the subnet. We must be careful making this change. Therefore I want the people mentioned above to raise their concerns.

qinqon reviewed (changes requested):
Are we sure it's ok to discard netmask ? it's important information, maybe we can make this as an opt-in feature.

qinqon commented:
@mfranczy the error you describe 
```
We treat VMs created by KubeVirt as worker nodes for another k8s cluster and during network initialization we realized that calico has a problem with issue described above, for instance:
```

Comes from a third party that you can change to discard netmask ? At the end maybe what is needed if to document the format reported by kubevirt, so third parties adapt to it, what do you think ? @phoracek @AlonaKaplan @EdDev @maiqueb

rmohr commented:
> Comes from a third party that you can change to discard netmask ? At the end maybe what is needed here is to document the format reported by kubevirt, so third parties adapt to it, what do you think ? @phoracek @AlonaKaplan @EdDev @maiqueb

Can be a solution. Somehow, calico in this case, gets access the the VM IP, so whatever gives access could probably also discard netmasks.

maiqueb commented:
I share the concern that this will potentially break all sorts of things that require a netmask.

> @mfranczy the error you describe
> 
> ```
> We treat VMs created by KubeVirt as worker nodes for another k8s cluster and during network initialization we realized that calico has a problem with issue described above, for instance:
> ```
> 
> Comes from a third party that you can change to discard netmask ? At the end maybe what is needed here is to document the format reported by kubevirt, so third parties adapt to it, what do you think ? @phoracek @AlonaKaplan @EdDev @maiqueb

This sounds like a good option: I feel it is the end consumer that knows what information it needs (compute the IP address from the provided CIDR).

EdDev commented on pkg/util/net/ip/ip.go:
```diff
@@ -19,10 +19,13 @@
 
 	netutils "k8s.io/utils/net"
+
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
```
This causes our general purpose ip package to depend on a specific virt-launcher API, which seems improper to me.
Generally speaking, this package should not depend on anything internal except what is under "pkg/util".

AlonaKaplan commented on pkg/virt-handler/vm.go:
```diff
@@ -605,6 +606,14 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 					newInterface.InterfaceName = interfaceStatus.InterfaceName
+
+					// Qemu guest reports CIDR, only an IP address should be passed to the status
+					err := iputils.ChangeCIDRsToIPs(&interfaceStatus)
```
First of all that a lot for finding the issue and the contribution.

1. I find it confusing changing the cidrs to ips and not just returning the ips.
2. I would call the method only if `!isForwardingBindingInterface`.
3. The method should be invoked in the [un-handled](https://github.com/kubevirt/kubevirt/blob/b561ed9ede0eac44ef3667c75ece78c9538ed1b8/pkg/virt-handler/vm.go#L621) domain interfaces loop as well.

Anyway,
I think the correct place for the change is the [agent_parser](https://github.com/kubevirt/kubevirt/blob/aa8a15d540ad90d7022be59b30840d41581f52bc/pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go#L268) and not here. IIUC, libvirt reports `ipAddr.IP` and `ipAddr.Prefix`. Our poller is covering it to 'ip\prefix`.
Since there may be users that are already using the cidr information I suggest that besides changing the poller `extractIPs` method, also adding the vmi api a new status field for the prefix.

AlonaKaplan replied:
@EdDev what do you think about the suggestion here?

EdDev replied:
I tend to agree that a better place to handle this is at the source, asking explicitly for the IP address (only) in the agent parser. The challenge here is how to keep both data available without duplicating code.

I am not raising here the issue about where and if we should expose the masks.

EdDev commented on pkg/util/net/ip/ip.go:
```diff
@@ -82,3 +85,28 @@ func isIPv6Disabled(filename string) bool {
+			return fmt.Errorf("qemu-guest-agent reported invalid CIDR value '%s' for '%s': %v",
+				ifs.Ip, ifs.InterfaceName, err)
+		}
+		ifs.Ip = ip.String()
```
Making this change mutable is not safe as it may change the values even if failure occurs later on, causing recovery to be questionable.
Try using a mutable implementation, returninig the data so the caller can update it (or not, in case of error).

EdDev commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1941,8 +1944,9 @@ var _ = Describe("VirtualMachineInstance", func() {
 					Mac:  mac,
-					Ip:   ip,
-					IPs:  ips,
+					// Qemu guest agent reports CIDR
```
This is specified in the variable names, unless there is a hidden added info here, I would suggest to drop the comment.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -605,6 +606,14 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 				if interfaceStatus, exists := domainInterfaceStatusByMac[interfaceMAC]; exists {
 					newInterface.InterfaceName = interfaceStatus.InterfaceName
+
+					// Qemu guest reports CIDR, only an IP address should be passed to the status
```
I would expect the name of the function to say this without the need to add comments here.

EdDev commented on pkg/util/net/ip/ip.go:
```diff
@@ -82,3 +85,28 @@ func isIPv6Disabled(filename string) bool {
 }
+
+func ChangeCIDRsToIPs(ifs *api.InterfaceStatus) error {
+	// if IP address was the passed value we should do nothing
```
I am not fun of comments. If possible, please express this in the func doctext (I am fine with just dropping as well).

EdDev reviewed (changes requested):
> So thinking about this, I am not sure this is good. I am not sure it is good to have APIs where people have to study manuals to see if in some corner-cases the format of values may change and break assumptions and software. How about splitting this in IPs and CIDRs? Both would contain the same IPs, except in one case the CIDR stripped. Exist there some corner-cases where this clarification would lead to unwanted side-effects from a networking perspective?

This seems to me no less confusing as in essence we will provide two API/s for almost the same data.
The Pod IP does not include the mask because the interest there is for accessing the Pod, and it did not account for multiple addresses in the same family on one or more interfaces.
On a VM, such cases may exist, therefore it makes sense to include the mask and it should be up to the users to realize the types are not the same (pod-ip vs vmi-ip). We can express this through the documentation to make it clearer as suggested.

As I see it, K8S API has an issue with backward compatibility and adding there more fields may merit the cost, but here we do not have such a problem and I do not see a reason to go a few steps back.

We can provide helpers to retrieve the needed info in the needed format or even ref the pod on which the info is available (maybe it is already in place), but as some suggested above, it is up to the caller to use the information as it is.
BTW, the name "ip" does not imply "address" nor "cidr", and in many systems using CIDR is the accepted standard.

mfranczy commented:
Thanks for review @EdDev @AlonaKaplan.

However, since you propose a bit diferent approaches, could you agree internally on some solution how this should be handled and write it here? Should I add a new field in the API or do you want to just express this in the doc and abandon this PR?

rmohr commented:
> This seems to me no less confusing as in essence we will provide two API/s for almost the same data.
> The Pod IP does not include the mask because the interest there is for accessing the Pod, and it did not account for multiple addresses in the same family on one or more interfaces.
> On a VM, such cases may exist, therefore it makes sense to include the mask and it should be up to the users to realize the types are not the same (pod-ip vs vmi-ip). We can express this through the documentation to make it clearer as suggested.
> 

I am not sure what the best solution is, but right now we mix `ip/mask` and `ip` in `IPs` which is probably the worst of all solutions. As we can for instance see in the gardener case and this PR, every consumer will have  to deal with probably even code-wise extra with this exception (`net.ParseIP` only works on `ip` and `net.ParseCIDR` only works on  `ip/mask`.

> As I see it, K8S API has an issue with backward compatibility and adding there more fields may merit the cost, but here we do not have such a problem and I do not see a reason to go a few steps back.

I am not sure what you mean with this statement. Do you mean just documenting it or somehow making it consistent while keeping the mask?

> 
> We can provide helpers to retrieve the needed info in the needed format or even ref the pod on which the info is available (maybe it is already in place), but as some suggested above, it is up to the caller to use the information as it is.
> BTW, the name "ip" does not imply "address" nor "cidr", and in many systems using CIDR is the accepted standard.

It is a pretty reasonable guess, that if we keep the reporting mixed, that people will assume that `IPs` is just `IPs` like in pods and services. Why would they ever look at the docs or look for helpers if they don't need CIDRs? We make it even worse because by default, in many of their labs and so on, they will only ever see a plain IP without masks most of the time It may even work on the customer side, but then suddenly it breaks.

Whatever we do, mixing the notation inside the field is from my perspective the worst we can do.

qinqon commented:
@mfranczy I have one question, looking at kubevirtci cluster (it's backed up by calico CNI) I don't see netmask at IP/IPs, what's the different with your deploy ?

EdDev commented:
> > This seems to me no less confusing as in essence we will provide two API/s for almost the same data.
> > The Pod IP does not include the mask because the interest there is for accessing the Pod, and it did not account for multiple addresses in the same family on one or more interfaces.
> > On a VM, such cases may exist, therefore it makes sense to include the mask and it should be up to the users to realize the types are not the same (pod-ip vs vmi-ip). We can express this through the documentation to make it clearer as suggested.
> 
> I am not sure what the best solution is, but right now we mix `ip/mask` and `ip` in `IPs` which is probably the worst of all solutions. As we can for instance see in the gardener case and this PR, every consumer will have to deal with probably even code-wise extra with this exception (`net.ParseIP` only works on `ip` and `net.ParseCIDR` only works on `ip/mask`.

For a VMI, I expect to report only CIDR formats in the IP/s fields.
While on the Pod, it is reported as IP addresses only.

Are we talking about the mix between the "primary" interface (pod network) and the rest? The first not including the mask and the rest including it? If that is the case, we have indeed a major problem.

> 
> > As I see it, K8S API has an issue with backward compatibility and adding there more fields may merit the cost, but here we do not have such a problem and I do not see a reason to go a few steps back.
> 
> I am not sure what you mean with this statement. Do you mean just documenting it or somehow making it consistent while keeping the mask?

Consistent keeping the mask and documenting it (if not already done).

> 
> > We can provide helpers to retrieve the needed info in the needed format or even ref the pod on which the info is available (maybe it is already in place), but as some suggested above, it is up to the caller to use the information as it is.
> > BTW, the name "ip" does not imply "address" nor "cidr", and in many systems using CIDR is the accepted standard.
> 
> It is a pretty reasonable guess, that if we keep the reporting mixed, that people will assume that `IPs` is just `IPs` like in pods and services. Why would they ever look at the docs or look for helpers if they don't need CIDRs? We make it even worse because by default, in many of their labs and so on, they will only ever see a plain IP without masks most of the time It may even work on the customer side, but then suddenly it breaks.
> 
> Whatever we do, mixing the notation inside the field is from my perspective the worst we can do.

I agree, we should have no mixing in the same field... that is pure wrong.
I expect all IP reports in the VMI to have CIDR, otherwise have nothing.

mfranczy commented:
@qinqon I am using multus and have installed qemu-guest-agent on VMIs, the agent reports CIDR. Look at https://github.com/kubevirt/kubevirt/pull/4217/files#diff-14837489e428608f8ed1f61f9d7242bcL1911 as it describes my scenario well.

rmohr commented:
> For a VMI, I expect to report only CIDR formats in the IP/s fields.
While on the Pod, it is reported as IP addresses only

I would say that this is true, as long as one uses multus to pass through a plain bridges and anything could happen. For the rest I am not that sure. If one uses only normal CNI plugins and/or masquerade, the netmask information has no more value than on pods (it may actually suggest to users that it has more value than it actually has).

> Are we talking about the mix between the "primary" interface (pod network) and the rest? The first not including the mask and the rest including it? If that is the case, we have indeed a major problem.
> 

Yes, by default one only sees a clean ip without submask. This leads to the situation that even if it works for you right now, as soon as one decides to start the first VM with a guest agent, entries with netmask get added.

EdDev commented:
> > For a VMI, I expect to report only CIDR formats in the IP/s fields.
> > While on the Pod, it is reported as IP addresses only
> 
> I would say that this is true, as long as one uses multus to pass through a plain bridges and anything could happen. For the rest I am not that sure. If one uses only normal CNI plugins and/or masquerade, the netmask information has no more value than on pods (it may actually suggest to users that it has more value than it actually has).

With multiple interfaces and multiple IP addresses, the mask has great importance to understand the network subnet.
It helps to understand if the configuration is not overlapping and what actually exists on the VM.

In my view, the VMI should report what it knows about the network, not what it calculated by heuristics.
It should always report its own addresses and not the one the world sees through nat/masquerade. If there is a need to do expose it through the VMI object, then there should be a dedicated "public" IP section which may fit to what you suggested (having two fields).

EdDev commented:
> If there is a need to do expose it through the VMI object, then there should be a dedicated "public" IP section which may fit to what you suggested (having two fields).

I guess a compromise could be to report only addresses through the existing fields and export the CIDR through a guest agent subsection like [this](https://kubevirt.io/api-reference/master/definitions.html#_v1_virtualmachineinstanceguestosinfo). (In our case, we can call it "guestInterfaceInfo" or something similar).
But I still feel uncomfortable with the notion that in a multi interface system we report the address without the mask, it can just make it very hard to understand if there are overlaps and alike.

In this case and for the pod network itself, we may have no choice anyway as this is a done deal in terms of API evolution, so my arguments may collide with current reality.
In any case, my 1st choice will be to always report with the mask/prefix. My 5th choice would be to report the CIDR ones in full under a guest-info subsection (we will have duplication with the 2nd option but at least will not have that ugly mix).
Please choose the 1st :)

mfranczy commented:
@EdDev @rmohr I will start working on that again next week.

> In any case, my 1st choice will be to always report with the mask/prefix. My 5th choice would be to report the CIDR ones in full under a guest-info subsection (we will have duplication with the 2nd option but at least will not have that ugly mix).
Please choose the 1st :)

I am going to follow the 1st option.

rmohr commented:
> I am going to follow the 1st option.

We discussed it in the community meeting. I think the api breakage with option 1 would be very big, because probably 95% of all kubevirt users have never seen an IP with the netmask. I would vote for separating them, especially since the netmask is for default CNI plugins normally not relevant. But we should definitely report also what the guest sees.

rmohr commented:
Ok, since no further comments came up, I would suggest to not add CIDRs to IPs in the `IPs` field (minimal API breakage) and instead either here or in a follow up PR add a `CIDRs` API field. Further we may want to consider, ass added by @EdDev, adding explicit entries in the guest agent structs to also "just" report unprocessed what the guest agent sees (local links, and so forth).

mfranczy commented:
Sorry for the delay with this. I had quite busy time.

@AlonaKaplan thanks for hint with the agent poller, I followed your idea.
@rmohr I can add a CIDR field in the following PRs to this and client-go repos.

Let's if any functional test will fail.

AlonaKaplan commented:
Thanks Marcin. The PR looks fine. I would really like the CIDRs field will be added to the status in this PR and not a follow-up one, since we may have users that need the cidr and now we just drop this information from the status.

AlonaKaplan commented on tests/vmi_multus_test.go:
```diff
@@ -296,7 +296,7 @@ var _ = Describe("[Serial]Multus", func() {
 
 				By("Configuring static IP address to ptp interface.")
-				Expect(configInterface(vmiOne, "eth0", "10.1.1.1/24")).To(Succeed())
+				Expect(configInterface(vmiOne, "eth0", "10.1.1.1")).To(Succeed())
```
I don't think you can drop the cidr from here.

mfranczy replied:
right, good point

AlonaKaplan commented on tests/vmi_multus_test.go:
```diff
@@ -499,8 +499,8 @@ var _ = Describe("[Serial]Multus", func() {
 					linuxBridgeNetwork,
 				}
 
-				ep1Ip := "1.0.0.10/24"
```
Same here, you cannot drop the CIDR. It is used for the `ip link add` command.

mfranczy replied:
I should be more careful while checking tests, thanks for catching this.

mfranczy replied:
Test is adjusted now, let's see how it goes.

mfranczy commented:
> Thanks Marcin. The PR looks fine. I would really like the CIDRs field will be added to the status in this PR and not a follow-up one, since we may have users that need the cidr and now we just drop this information from the status.

That's fair, thanks for input. Will do in this PR then.

mfranczy commented:
@AlonaKaplan I've added CIDR fields, I forgot that client-go is under `staging` dir so no need to open additional PR.

mfranczy commented:
/retest

AlonaKaplan commented on pkg/virt-launcher/virtwrap/agent-poller/agent_parser.go:
```diff
@@ -254,31 +254,53 @@ func convertInterfaceStatusesFromAgentJSON(agentResult []Interface) []api.Interf
-		interfaceIP = interfaceIPs[0]
+
+	// If ipv4 interface was found, set the IP as the first element.
+	// The first element will be set as a primary address.
```
Just a nit. I don't like the fact you specify here how the method should be used. How can you assure the first element will be used as the primary address? I prefer the method will return - ip, ips, cidr, cidrs.

mfranczy replied:
as you wish, I will change that

mfranczy replied:
> I don't like the fact you specify here how the method should be used. How can you assure the first element will be used as the primary address?

And.. just for clarity and explanation. I think I don't force anybody to use this method in a different way that was done before. The overall result is still the same. It's just a matter of returning primary address as a first element of list (which is assured by prepend) or returning it as a new value, but I think I get your point as returning a new value is explicit.

AlonaKaplan commented on api/api-rule-violations-known.list:
```diff
@@ -162,6 +162,7 @@ API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachin
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceList,Items
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceMigrationList,Items
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceMigrationStatus,Conditions
+API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceNetworkInterface,CIDRs
```
what about the CIDR?

mfranczy replied:
CIDR is not a list type `list_type_missing`.

AlonaKaplan commented on api/api-rule-violations-known.list:
```diff
@@ -251,6 +252,7 @@ API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInsta
 API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceGuestAgentInfo,GAVersion
 API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceGuestOSInfo,VersionID
 API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceMigrationState,MigrationUID
+API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceNetworkInterface,CIDRs
```
same

mfranczy replied:
`CIDR 'json:"cidr"'` doesn't cause API rule violation.

AlonaKaplan commented on api/api-rule-violations.list:
```diff
@@ -162,6 +162,7 @@ API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachin
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceList,Items
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceMigrationList,Items
 API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceMigrationStatus,Conditions
+API rule violation: list_type_missing,kubevirt.io/client-go/api/v1,VirtualMachineInstanceNetworkInterface,CIDRs
```
same

mfranczy replied:
this generated file, output from `make generate`.

AlonaKaplan commented on api/api-rule-violations.list:
```diff
@@ -251,6 +252,7 @@ API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInsta
 API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceGuestAgentInfo,GAVersion
 API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceGuestOSInfo,VersionID
 API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceMigrationState,MigrationUID
+API rule violation: names_match,kubevirt.io/client-go/api/v1,VirtualMachineInstanceNetworkInterface,CIDRs
```
same

mfranczy replied:
this generated file, output from `make generate`

AlonaKaplan commented on pkg/virt-handler/vm.go:
```diff
@@ -482,7 +482,6 @@ func (d *VirtualMachineController) getPodInterfacefromFileCache(uid types.UID, i
 
 func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstance, domain *api.Domain, syncError error) (err error) {
 	condManager := controller.NewVirtualMachineInstanceConditionManager()
-
```
Why did you remove this empty line?

mfranczy replied:
by accident sorry, will fix.

AlonaKaplan commented on pkg/virt-handler/vm.go:
```diff
@@ -579,7 +578,6 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 					// Only interfaces defined in domain.Spec are handled here
 					newInterface = existingInterface
 					newInterface.MAC = interfaceMAC
-
```
same

mfranczy replied:
by accident sorry, will fix.

AlonaKaplan commented:
/approve
@EdDev please review

mfranczy commented:
@AlonaKaplan thanks for another review round, I see multus test is failing and I have to discover if this is a test problem or something is wrong with logic.

mfranczy commented:
/hold

mfranczy commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1923,8 +1923,8 @@ var _ = Describe("VirtualMachineInstance", func() {
-			ip := "2.2.2.2/24"
-			ips := []string{"2.2.2.2/24", "3.3.3.3/16"}
+			ip := "2.2.2.2"
+			ips := []string{"2.2.2.2", "3.3.3.3"}
```
right, fixed.

EdDev reviewed (changes requested):
Thank you for this change.

This PR includes two changes, one I think we agreed on and the 2nd needs some discussion.
- Removing the CIDR from the VMI interface status has been agreed and reasoned with the wish to be in sync with the Pod IP status.
- Add a new entry that includes the CIDR.

First, I suggest to split these into two commits so we can continue the discussion on the second point while agreeing with the first.
Also, please elaborate as you did in the PR description into the commit message so we will have this information in the git log for future us.

The introduction of the CIDR, as I remember from the community meeting, has been discussed to be exposed through a general Guest Agent specific subtree. There we can include all available GA info.
This opens question as:
- Should it be a row json or should we structure it with all possible information?
- Should we add (as this current change suggests) an additional field with CIDR or just a `prefix` one? Will that be consistent with reports when there is no GA?
- How the new GA branch will be called? `guestagentdata`? `guestagentrawdata` ?

I suggest presenting all options (I would suggest with examples) in the community meeting or in the mailing list so an agreement can be reached.

Personally I would prefer for the API to be consistent and fields will not pop up under interfaces status based on the existence of GA or not. It requires users of this API to check a specific field (e.g. CIDR, prefix) for existance and to me it is a smell that something is not right. If we dedicate a specific subtree/entry for GA, that is easier to reason when it does not show.

Hope I did not complected the discussion too much.

EdDev commented:
Please also note that the "release note" in the PR description needs to be populated because this change is touching the API. It requires a short description of what changed.

AlonaKaplan commented:
As Eddy mentioned, please update the release note.
Also, once the PR is merged. Please send a PR to update the [documentation](https://github.com/kubevirt/user-guide/blob/master/usage/guest-agent.md).

@EdDev  regarding the "Guest Agent specific subtree". I believe the correct place of the CIDR is in the interface status, next to the ip.
As far as I understand from @rmohr  [comment](https://github.com/kubevirt/kubevirt/pull/4217#issuecomment-707126962), the GA struct is not instead of the status but an addition.
Regarding having both `CIDR` and `CIDRs`, I don't mind keeping just `CIDRs`.

EdDev commented:
> @EdDev regarding the "Guest Agent specific subtree". I believe the correct place of the CIDR is in the interface status, next to the ip.
> As far as I understand from @rmohr [comment](https://github.com/kubevirt/kubevirt/pull/4217#issuecomment-707126962), the GA struct is not instead of the status but an addition.

IMO it is not a fit for a proper API as it will not be there when the GA is not active and it may or may not be there due to other reasons (pod network vs secondary networks?).
This moves the need for handling this logic to the user of these fields and I think the fields should not "blink". This is always a problem with inconsistent reporting/api.

It is also a duplicate data, we only need the mask/prefix, not the full CIDR in this case. But this stands independently from the previous point.

EdDev commented:
This change was discussed in the last [community meeting](https://docs.google.com/document/d/1kyhpWlEPzZtQJSjJlAqhPcn3t0Mt_o0amhpuNPGs1Ls) (2020-11-11).

The option to provide the Guest Agent information in a raw format (json?) under a dedicated field has been favored.
As an AI, I suggest first to split the agreed change of removing the prefix from the IP fields in a separate commit. It may be easier to push this forward if it will be a dedicated PR. @AlonaKaplan WDYT?

The seconds step is to create a PR/commit that introduces a Guest Agent subtree in the VMI status and adds there relevant information. I guess the fields under that subtree are the different commands we execute to fetch the data.

rmohr commented:
> The option to provide the Guest Agent information in a raw format (json?) under a dedicated field has been favored.

We don't expos anything raw in the API and I don't understand why this can't be structured. I doubt that the guest-agent has an undefined API.

rmohr commented:
> We don't expos anything raw in the API and I don't understand why this can't be structured. I doubt that the guest-agent has an undefined API.

Also note that we have already guest-agent subresources where one can retrieve data which is too big to be part of the API (like installed packages and such).

EdDev commented:
> We don't expos anything raw in the API and I don't understand why this can't be structured. I doubt that the guest-agent has an undefined API.

I suspected this may pop up, that is why I suggested to move it to a different PR for discussion.
The GA info arrives to us in a JSON based format per what I see. The structure mentioned here is about if we want to define it in the spec explicitly or just dump the json as is under a string. The later has the advantage of not requiring us to depend on GA format in case it changes with each version (adding fields most likely).
Anyway, this discussion fits a dedicated PR IMO.

EdDev commented:
> Also note that we have already guest-agent subresources where one can retrieve data which is too big to be part of the API (like installed packages and such).

I would start with selected agree data first, we do not need to serve it all. But structuring the info in the spec is a valid option, I think it needs to be evaluated as well.

rmohr commented:
> I suspected this may pop up, that is why I suggested to move it to a different PR for discussion.
> The GA info arrives to us in a JSON based format per what I see. The structure mentioned here is about if we want to define it in the spec explicitly or just dump the json as is under a string. The later has the advantage of not requiring us to depend on GA format in case it changes with each version (adding fields most likely).
> Anyway, this discussion fits a dedicated PR IMO.

Unknown data dumps will never happen for obvious reasons: We have an object limit on k8s objects, list/watch transfers the whole object for any change which introduces spikes). Also like we see in this PR where we did not review things carefully enough, unexpected things can be reported. Therefore field needs to be carefully reviewed and nothing can just be exposed from an unknown sources and some info needs to go to subresources  and provided on-demand.

AlonaKaplan commented:
/approve cancel
@mfranczy I'm ok with the PR as is. But since the community prefers the cidr to be part of a dedicated guest-agent sub tree I'm removing the approve.

mfranczy commented:
@AlonaKaplan ok, I will deal with the PR again later this week, sorry I have to manage my time between projects.

mfranczy commented:
/hold cancel

mfranczy commented:
> Please also note that the "release note" in the PR description needs to be populated because this change is touching the API. It requires a short description of what changed.

done

> Anyway, this discussion fits a dedicated PR IMO.

@AlonaKaplan  @EdDev  @rmohr I reverted changes to minimal. If you don't mind let's continue the discussion about CIDR in some issue or I can open another PR with the commit were I introduced the CIDR field.

EdDev reviewed (commented):
Thank you.

Can you please update the commit message with all the info from the PR description?
It is beneficial for the info to live in the git logs so we can check the history later on.

Also, please try not to pass 72 chars or so as recommended [here](https://chris.beams.io/posts/git-commit/#seven-rules).
A suggested subject line can be: `virt-handler, status: Do not include the IP prefix consistently`
But anything goes, just try to make it a bit shorter.

mfranczy commented:
@EdDev I've updated the commit message as suggested, additionaly I had to change small thing at [libnet/cloud_init](https://github.com/kubevirt/kubevirt/pull/4217/files#diff-b03439ba2dac10103df177f90802bbfa4d14104b15a9247fa885e7d6d9bea91eR62) as integration tests were failing:

```
tests/network/primary_pod_network.go:107
Timed out after 5.000s.
should contain at IPs the IP and the cloud-init configured ipv6
Expected
    <[]string | len:3, cap:4>: [
        "10.244.140.91",
        "fd10:0:2::2",
        "fe80::c2b:aaff:fe84:1831",
    ]
to contain elements
    <[]interface {} | len:2, cap:2>: ["10.244.140.91", "fd10:0:2::2/120"]
the missing elements were
    <[]interface {} | len:1, cap:1>: ["fd10:0:2::2/120"]
tests/network/primary_pod_network.go:110
```

I run this locally after change at it passed. Let's see if this time tests will be green.

EdDev reviewed (approved):
Thank you!

rmohr commented:
/approve
